main.cpp

```diff
@@ -20,19 +21,39 @@ struct Wall
 {
     plane p;
     color c;
+
+    friend std::ostream& operator<<(std::ostream& os, const Wall&wall) {
+		return os << "Wall{ p" << wall.p << " c" << wall.c << "}";
+	 }
 };
 
 struct Sphere
 {
     vec3<float> center;
     float radius;
+
+    friend std::ostream& operator<<(std::ostream& os, const Sphere&sphere) {
+		return os << "Sphere{ center " << sphere.center << " radius " << sphere.radius << "}";
+	 }
 };
 
 struct Scene
 {
     vec3<float> eye;
     std::vector<Sphere> spheres;
     std::vector<Wall> walls;
+
+    friend std::ostream& operator<<(std::ostream& os, const Scene&scene) {
+		os << "Scene{" << std::endl
+		   << "  Eye:" << scene.eye << std::endl;
+		std::for_each(scene.spheres.begin(), scene.spheres.end(), [](Sphere const& sphere) {
```

Comment:
Honestly, I never understood the point of `std::for_each` in this particular context. Something like

```c++
for (const auto &sphere: spheres) {
    std::cout << " " << sphere << std::endl;
}
```

is

1. Way more readable
2. Probably produces faster code

The only case where I see `std::for_each` applicable is when it's passed as an argument to a higher order functions. But this is not the case.

Please replace all of the `std::for_each` instances with the foreach syntax of C++11